packages/zoltan2/src/algorithms/color/Zoltan2_AlgHybridGMB.hpp

```diff
@@ -0,0 +1,458 @@
+    typedef typename Adapter::gno_t gno_t;
+    typedef typename Adapter::offset_t offset_t;
+    typedef typename Adapter::scalar_t scalar_t;
+    typedef typename Adapter::base_adapter_t base_adapter_t;
+    typedef Tpetra::Map<lno_t, gno_t> map_t;
+    typedef int femv_scalar_t;
+    typedef Tpetra::FEMultiVector<femv_scalar_t, lno_t, gno_t> femv_t;
+    
+  private:
+
+    void buildModel(modelFlag_t &flags); 
+
+    //function to invoke KokkosKernels distance-1 coloring    
+    template <class ExecutionSpace, typename TempMemorySpace, 
+              typename MemorySpace>
+    void colorInterior(const size_t nVtx, 
+                       Kokkos::View<lno_t*, Kokkos::Device<ExecutionSpace,MemorySpace> > adjs_view,
+                       Kokkos::View<offset_t*, Kokkos::Device<ExecutionSpace,MemorySpace> > offset_view, 
+                       Teuchos::ArrayRCP<int> colors,
+                       Teuchos::RCP<femv_t> femv,
+                       bool recolor=false){
+      
+      //default values are taken from KokkosKernels_TestParameters.hpp
+      
+      int algorithm = pl->get<int>("Hybrid_algorithm",0);
+      int repeat = pl->get<int>("Hybrid_repeat",1); //probably not using this 
+                                                    //until we test performance
+      int chunk_size = pl->get<int>("Hybrid_chunk_size",-1); 
+      int shmemsize = pl->get<int>("Hybrid_shmemsize", 16128);
+      int team_size = pl->get<int>("Hybrid_team_size", -1);
+      int use_dynamic_scheduling = pl->get<int>("Hybrid_use_dynamic_scheduling",
+                                                0);
+      int verbose = pl->get<int>("Hybrid_verbose",0);
+
+      int vector_size = pl->get<int>("Hybrid_vector_size",-1);
+
+      typedef KokkosKernels::Experimental::KokkosKernelsHandle
+          <size_t, lno_t, lno_t, ExecutionSpace, TempMemorySpace, 
+           MemorySpace> KernelHandle;
+      KernelHandle kh;
+
+      kh.set_team_work_size(chunk_size);
+      kh.set_shmem_size(shmemsize);
+      kh.set_suggested_team_size(team_size);
+      kh.set_suggested_vector_size(vector_size);
+       
+      kh.set_dynamic_scheduling(use_dynamic_scheduling);
+      kh.set_verbose(verbose);
+    
+      if(recolor){
+        kh.create_graph_coloring_handle(KokkosGraph::COLORING_VBBIT);
+      } else {
+        kh.create_graph_coloring_handle(KokkosGraph::COLORING_DEFAULT);
+        
+      }
+      //set the initial coloring of the kh.get_graph_coloring_handle() to be
+      //the data view from the femv.
+      Kokkos::View<int**, Kokkos::LayoutLeft> femvColors = femv->template getLocalView<MemorySpace>();
+      Kokkos::View<int*, Tpetra::Map<>::device_type >  sv = subview(femvColors, Kokkos::ALL, 0);
+      kh.get_graph_coloring_handle()->set_vertex_colors(sv);
+      
+      KokkosGraph::Experimental::graph_color_symbolic(&kh, nVtx, nVtx, 
+                                                      offset_view, adjs_view);
+      
+      Kokkos::fence();
```

Comment:
Why do you need a fence here?

Reply:
I don't think I do. My GPU-based platform ran tests just fine without it, so I have removed this fence.